plugins/test/framework.cc

```diff
@@ -18,13 +18,53 @@
 
 namespace service_extensions_samples {
 
+size_t Buffer::size() const {
+  if (string_buffer_) {
+    return string_buffer_->length();
+  }
+  return proxy_wasm::BufferBase::size();
+}
+proxy_wasm::WasmResult Buffer::copyTo(proxy_wasm::WasmBase* wasm, size_t start,
+                                      size_t length, uint64_t ptr_ptr,
+                                      uint64_t size_ptr) const {
+  if (string_buffer_) {
+    auto buffer_section = string_buffer_->substr(start, length);
+    std::string_view s(buffer_section);
+    if (!wasm->copyToPointerSize(s, ptr_ptr, size_ptr)) {
+      return proxy_wasm::WasmResult::InvalidMemoryAccess;
+    }
+    return proxy_wasm::WasmResult::Ok;
+  }
+  return proxy_wasm::BufferBase::copyTo(wasm, start, length, ptr_ptr, size_ptr);
+}
+
+proxy_wasm::WasmResult Buffer::copyFrom(size_t start, size_t length,
+                                        std::string_view data) {
+  if (string_buffer_) {
+    if (start == 0) {
+      if (length != 0) {
+        string_buffer_->erase(0, length);
+      }
+      string_buffer_->insert(start, data);
```

Comment:
AIUI this could be quite an expensive operation on std::string since it could shift/copy all of the std::string contents? I think we should compose a series of string_views here. When wasm publishes a new string, we take local ownership, but we leave alone any bits of the body that aren't being touched.

As a (complex) example, Envoy's Buffer implementation stores a collection of slices to handle data inserts at the front and back: https://github.com/envoyproxy/envoy/blob/ad0a10013c977a0f6a1e9de876d0a9eb22f1a187/source/common/buffer/buffer_impl.h#L773

We don't have to be quite as optimized, but we should avoid copying any content that the wasm doesn't mutate. My other comment about reducing string copies implies that ideally the Buffer should hold a mix of externally owned strings (the input) and internally owned mutations (received via setBuffer).

Reply:
I'd propose that for this stage, we start with a basic copy-on-write Buffer impl (which BufferBase provides 90% of), then change to something Cord-like as a further refinement later on. Just to get something simple and "good enough" landed for now.